### DISCUSSION TITLE
How to set default device as CPU

### DISCUSSION THREAD

**Discussion by zohimchandani** [jax-ml/jax#10399] [Q&A]
I would like to benchmark the performance of my code on CPU and GPU but Jax natively recognizes the GPU. How can I force it to run the code on CPU?

**Comment by YouJiacheng** [comment]
Option 1
```python
import jax
import os
os.environ['CUDA_VISIBLE_DEVICES'] = ''
# before execute any computation / allocation
print(jax.numpy.ones(3).device()) # TFRT_CPU_0
```
Option 2
```python
import jax
jax.config.update('jax_platform_name', 'cpu')
# before execute any computation / allocation
print(jax.numpy.ones(3).device()) # TFRT_CPU_0
```
Option 3
```python
import jax

def f():
    return jax.numpy.ones(3)

print(jax.jit(f, backend='cpu')().device()) # TFRT_CPU_0
```

**Reply by zohimchandani** [reply]
Trying option 2: 
<img width="612" alt="Screenshot 2022-04-21 at 13 50 01" src="https://user-images.githubusercontent.com/60148725/164461603-f960bdf5-edf6-4bc8-b9ff-622d815257c1.png">

**Reply by jakevdp** [reply]
Note that this setting will only work if it is run before any other JAX code in the session – the device cannot be changed via configuration flags mid-session. If you're using Jupyter, try restarting your kernel and then rerun the code.

**Reply by zohimchandani** [reply]
@jakevdp 
Perfect, got it, any ideas what `TFRT` stands for?

Also, given that my remote machine has 20 CPUs, can `pmap` be used to parallelize the compuation over the CPUs or does it only support GPUs?

**Reply by YouJiacheng** [reply]
I think it is https://github.com/tensorflow/runtime/blob/master/documents/tfrt_host_runtime_design.md.
There is a config option to split the `TFRT_CPU_0`, which already represent all CPU cores, into multiple device for debugging purpose. But the performance is inferior.
```python
os.environ['XLA_FLAGS'] = '--xla_force_host_platform_device_count=8'
```
